File: testdata/fixtures/kotlin/CWE-532.kt
// CWE-532: a secret is written to the log.
//
// Android's log is a ring buffer, and everything an app writes to it goes into
// the bug reports the user sends, the crash reports the vendor collects, and -
// on a rooted or debuggable device, or through any app holding READ_LOGS - the
// screen of whoever asked. A password or a token in there is a password or a
// token in a file the app does not control the lifetime of.
//
// The interesting distinction is between logging a secret and logging *about*
// one. `Log.d(TAG, "password rejected")` is a message; `Log.d(TAG, password)`
// is the password.
package fixture

import android.util.Log

class Session {

    fun login(password: String, token: String) {
        // ruleid: kotlin-sensitive-value-logged
        Log.d(TAG, "authenticating with password $password")

        // ruleid: kotlin-sensitive-value-logged
        Log.i(TAG, "token=" + token)

        // ruleid: kotlin-sensitive-value-logged
        Log.v(TAG, apiSecret)
    }

    fun report(user: String) {
        // ok: kotlin-sensitive-value-logged
        Log.w(TAG, "password rejected for this account")

        // ok: kotlin-sensitive-value-logged
        Log.d(TAG, "signed in as $user")

        // ok: kotlin-sensitive-value-logged
        Log.e(TAG, "could not reach the token endpoint")
    }

    private val apiSecret = System.getenv("API_SECRET")

    companion object {
        private const val TAG = "Session"
    }
}

File: testdata/fixtures/swift/CWE-532.swift
// CWE-532: the secret ends up in a log the app does not own.
//
// `print` and `NSLog` write to the unified log, which is not a developer-only
// place: it is collected into sysdiagnose bundles the user is asked to send to
// support, it is visible in Console.app to anyone with the phone plugged in
// and trusted, and on a shared Mac it is readable in the simulator's log by
// every account. A token written there has a lifetime the app does not
// control.
//
// The whole difficulty is telling a secret from a message about one.
// `print("password rejected")` is a sentence with the word in it;
// `print(password)` is the password. What separates them is that the second
// carries a *value* - a bare name, or a `\(...)` interpolation - and the first
// is only letters.
import Foundation
import os

func audit(user: String, password: String, sessionToken: String) {

    // ruleid: swift-sensitive-value-logged
    print(password)

    // ruleid: swift-sensitive-value-logged
    print("auth token: \(sessionToken)")

    // ruleid: swift-sensitive-value-logged
    NSLog("password for %@ is \(password)", user)

    // ruleid: swift-sensitive-value-logged
    debugPrint(sessionToken)

    // ok: swift-sensitive-value-logged
    print("password rejected")

    // ok: swift-sensitive-value-logged
    print("signing in \(user)")

    // ok: swift-sensitive-value-logged
    NSLog("token endpoint unreachable")

    // ok: swift-sensitive-value-logged
    print(user)
}
